fix: Resolve symlink before FlakeRefFromEnv

Darkness9724 opened the pull request:
Greetings!

Currently, `nixos-cli` subcommands that execute the `nix` command under the hood fail to [find flake.nix inside symlinked directories](https://github.com/NixOS/nix/issues/8013) (for example, `ln -s ~/.config/nixos /etc; nixos apply -dv`). However, `nixos-rebuild{,-ng}` work fine in the same setup.

This change ensures that the path is resolved to the real directory before being passed to `FlakeRefFromEnv` using `filepath.EvalSymlinks`, allowing symlinked configuration directories to work transparently.

water-sucks reviewed (changes requested):
Thanks for the PR! I was not aware of this issue.

This PR currently only does this for the generic `FindConfiguration()` function, which only resolves the issue for implicit configuration discovery.

However, many commands accept explicit command-line flake ref parameters, so this change may need to be applied to all those places, or the same problem may arise.

In our case, since this is flake-specific behavior, I think it would make sense to move this to `internal/configuration/configuration.go`, and only evaluate symlinks if the provided path exists on disk (i.e. verified with `os.Stat()`). I'm willing to hear alternative ways of implementing this though :}

Darkness9724 committed:
fix: Resolve symlink before FlakeRefFromEnv

Currently, nixos-cli subcommands that execute the nix command under the hood
fail to find flake.nix inside symlinked directories (ex: ln -s ~/.config/nixos /etc; nixos apply -dv).
However, nixos-rebuild{,-ng} work fine in the same setup.

This change ensures that the path is resolved to the real directory
before being passed to FlakeRefFromEnv using filepath.EvalSymlinks,
allowing symlinked configuration directories to work transparently.

Signed-off-by: Maxim Mikhailov <darkness9724@gmail.com>

Darkness9724 commented:
> In our case, since this is flake-specific behavior, I think it would make sense to move this to `internal/configuration/configuration.go`, and only evaluate symlinks if the provided path exists on disk (i.e. verified with os.Stat()). I'm willing to hear alternative ways of implementing this though :}

I believe you meant `internal/configuration/flake.go`? Because the previous change was already applied to `internal/configuration/configuration.go`.

I’ve gone ahead and moved the symlink resolution logic to `internal/configuration/flake.go` as suggested. The new `ResolvePath` function uses `os.Stat()` to verify path existence before calling `filepath.EvalSymlinks`. It’s now should be integrated into both `FlakeRefFromEnv` for implicit configuration discovery and `FlakeRefFromString` for explicit flake references.

I’ve also tested cases like `nixos apply git+https://path/to/repo/` to ensure non-local URIs are handled correctly. I wanted to add unit tests as well, but I’m not very experienced with Go yet :)

water-sucks commented on internal/configuration/flake.go:
```diff
@@ -19,18 +20,46 @@ type FlakeRef struct {
+			uri = resolved
+		}
+	}
+
```
IMO we can probably just attempt to call os.Stat(), and if it fails, just use the URI as is. It's highly unlikely that someone is going to have a path name with `://` in the name anyway.

Open to suggestions though!

```suggestion
	if _, err := os.Stat(path); err == nil {
		if resolved, err := filepath.EvalSymlinks(path); err == nil {
			uri = resolved
		}
        }
```

water-sucks commented on internal/configuration/flake.go:
```diff
@@ -45,7 +74,12 @@ func FlakeRefFromEnv(defaultLocation string) (*FlakeRef, error) {
+		return nil, fmt.Errorf("failed to resolve path %s: %w", nixosConfig, err)
+	}
+
+	return FlakeRefFromString(resolvedConfig), nil
```
This repeats the resolve path logic twice, once in the `FlakeRefFromString()` function, so the original code suffices in this case.

water-sucks commented on internal/configuration/flake.go:
```diff
@@ -19,18 +20,46 @@ type FlakeRef struct {
+
+	return resolved, nil
+}
+
```
If the previous changes are applied, this function is no longer needed.

```suggestion
```

Darkness9724 replied:
Yeah, you're right. This is indeed a way simpler, straighter approach than mine. Tested it on typical use cases, and it does the same thing.
Thanks, I’ve pushed the revised commit.

water-sucks reviewed (approved):
Thanks for the contribution! Works during my own (rudimentary) testing as well.

Appreciate you for taking the time to address the requested changes too :}

At some point I want to take a crack at resolving symlinks in upstream Nix as well, since that can be useful for behavior for other systems, i.e. `nix-darwin`, but this will do for now.
